#189: Configure column name converter for auto schema inference (#190)
exasol/virtual-schema-common-document: doc/user_guide/edml_user_guide.md

  "addSourceReferenceColumn": true
}
```

You can use this property for all dialects. Typically, it will, however, only give you additional information, if you load data from multiple sources.

The `SOURCE_REFERENCE` column has a maximum size of 2000 characters. The adapter will throw an exception when a source reference exceeds this.

## Define Mapping of Fields to Table Columns

EDML allows you to specify custom mapping from fields in source files to Exasol table columns. Virtual schema support two options for the mapping:
* [Automatic Mapping Inference](#automatic-mapping-inference)
* [Explicit Mapping Definition](#explicit-mapping-definition)

Automatic mapping inference is only supported for [file based virtual schemas](https://github.com/exasol/virtual-schema-common-document-files) using Parquet and CSV files. All other virtual schemas and file formats require an explicit mapping definition.

### Automatic Mapping Inference

To use automatic mapping inference, just omit the `mapping` element from the EDML definition. The virtual schema will then infer the mapping from the schema of the source. Currently this is only supported for Parquet and CSV files.

#### Notes

> * The files specified in the `source` must be available when creating the virtual schema. If the files are not available, the `CREATE VIRTUAL SCHEMA` command will fail.
>   * When you don't use automatic mapping inference (i.e. you specify the `mapping` element) you can still create the virtual schema as before without `source` files being available.

Reviewer: ```suggestion
* When specifying an explicit mapping in property `mapping` then statement `CREATE VIRTUAL SCHEMA` will even be successful if property `source` is missing,
* However, when using automatic mapping inference and property `source` is missing, then statement `CREATE VIRTUAL SCHEMA` will fail.
```

> * The adapter will detect the mapping based on the schema of the first file. Please make sure that all files specified as `source` are using the same schema, else the mapping may be wrong.

Reviewer: Does this also apply if using multiple mappings / files / entries each with potentially different source?

Author: You can specify only a single source, that potentially matches multiple files, e.g. using a wildcard expression.
If you want to have multiple sources, you must map them to separate tables.

* The adapter will detect the mapping when the virtual schema is created. If the schema of the `source` files changes, please drop and re-create the virtual schema to run the auto-inference again.

> * Creating the virtual schema with auto-inference will take longer because the adapter needs to read files from the `source`.

Reviewer: ```suggestion
* Creating the virtual schema with auto-inference will take longer because the adapter needs to read files from the `source` in order to infer the mapping.
```

* Please see [below](#automatic-mapping-inference-for-csv-files) for details about auto-inference for CSV files.

#### Column Name Conversion

By default the virtual schema will convert source column names to `UPPER_SNAKE_CASE` for Exasol column names during automatic mapping inference. If you want to use the original name from the source file, you can add property `autoInferenceColumnNames` to the EDML definition. This property supports the following values:
* `CONVERT_TO_UPPER_SNAKE_CASE`: Convert column names to `UPPER_SNAKE_CASE` (default).

> * `KEEP_ORIGINAL_NAME`: Do not convert column names, use column name from source.

Reviewer: ```suggestion
* `KEEP_ORIGINAL_NAME`: Do not convert column names, use the column names as specified in the data source.
```

Example:

```json
{
  "$schema": "https://schemas.exasol.com/edml-2.1.0.json",
  "source": "data/CsvWithHeaders.csv",
  "destinationTable": "BOOKS",
  "autoInferenceColumnNames": "KEEP_ORIGINAL_NAME"
}
```

> Exasol identifiers like column names must conform to certain criteria, see the [SQL Identifier documentation](https://docs.exasol.com/db/latest/sql_references/basiclanguageelements.htm#SQLIdentifier) for details. If the column names in your source files are invalid Exasol SQL identifiers, the mapping or queries may fail. In this case we recommend using option `CONVERT_TO_UPPER_SNAKE_CASE`.

Reviewer: ```suggestion
The column names must be valid [Exasol SQL Identifiers](https://docs.exasol.com/db/latest/sql_references/basiclanguageelements.htm#SQLIdentifier).  If the column names in your source files are invalid, the mapping or queries may fail. In this case we recommend using option `CONVERT_TO_UPPER_SNAKE_CASE`.
```

The nested property `author.name` is mapped to the column `AUTHOR_NAME`.

> In order to let the adapter create the described mapping we create the following definition in the `mapping` property:

Reviewer: ```suggestion
In order to let the adapter create the described mapping we create the following definition in property `mapping`:
```

The Virtual Schema adapter automatically adds a foreign key to the table. In the example above, it adds the column `BOOKS_ISBN` to the `BOOKS_TOPICS` table. It did pick the `ISBN` column, because we marked it as a key column.

#### Key Types

> There are two different key types: `global` and `local`. The difference only plays a role when mapping multi-level nested lists.

Reviewer: ```suggestion
You can map multi-level nested lists with two different key types: `global` and `local`.
```
